File: examples/cpp/06_playback_and_delete.cpp
#include <filesystem>
#include <iostream>

#include "example_utils.h"

int main(int argc, char** argv) {
    audio_sdk::AudioManager manager;

    manager.SetEventCallback([](const audio_sdk::Event& event) {
        std::cout << "event=" << audio_sdk_examples::EventTypeName(event.type)
                  << " device=" << event.device_id
                  << " details=\"" << event.details << "\"\n";
    });

    if (!audio_sdk_examples::SelectFirstDevice(manager, audio_sdk::DeviceDirection::kInput)) {
        return 1;
    }
    if (!audio_sdk_examples::SelectFirstDevice(manager, audio_sdk::DeviceDirection::kOutput)) {
        return 1;
    }

    const std::filesystem::path path =
        argc > 1 ? std::filesystem::path(argv[1]) : audio_sdk_examples::RecordingPath("cpp_playback_example.wav");

    if (!std::filesystem::exists(path)) {
        std::cout << "No existing WAV supplied, recording a short clip first: " << path << '\n';
        if (!audio_sdk_examples::Require(manager.StartRecording(path.string()), "start recording")) {
            return 1;
        }
        audio_sdk_examples::SleepForRecording();
        if (!audio_sdk_examples::Require(manager.StopRecording(), "stop recording")) {
            return 1;
        }
    }

    if (!audio_sdk_examples::Require(manager.PlayRecording(path.string()), "play recording")) {
        return 1;
    }

    if (!audio_sdk_examples::Require(manager.DeleteRecording(path.string()), "delete recording")) {
        return 1;
    }

    std::cout << "Played and deleted " << path << '\n';
    return 0;
}

File: examples/python/06_playback_and_delete.py
import sys
from pathlib import Path

import audio_sdk


def main() -> None:
    manager = audio_sdk.AudioManager()

    def on_event(event: object) -> None:
        print(
            f"event={audio_sdk.event_type_name(event.type)} "
            f"device={event.device_id} "
            f'details="{event.details}"'
        )

    manager.set_event_callback(on_event)
    input_device = audio_sdk.select_first_device(manager, audio_sdk.DeviceDirection.INPUT)
    output_device = audio_sdk.select_first_device(manager, audio_sdk.DeviceDirection.OUTPUT)
    print(f"Selected input device: {input_device.name} ({input_device.id})")
    print(f"Selected output device: {output_device.name} ({output_device.id})")

    path = Path(sys.argv[1]) if len(sys.argv) > 1 else audio_sdk.recording_path("python_playback_example.wav")
    if not path.exists():
        print(f"No existing WAV supplied, recording a short clip first: {path}")
        audio_sdk.require(manager.start_recording(str(path)), "start recording")
        audio_sdk.sleep_for_recording()
        audio_sdk.require(manager.stop_recording(), "stop recording")

    audio_sdk.require(manager.play_recording(str(path)), "play recording")
    audio_sdk.require(manager.delete_recording(str(path)), "delete recording")
    print(f"Played and deleted {path}")


if __name__ == "__main__":
    main()
